euphoria-hadoop/src/main/java/cz/seznam/euphoria/hadoop/output/SequenceFileSink.java

```diff
@@ -52,263 +51,143 @@
-
-    /**
-     * Optional setter if not used it will be created new hadoop configuration.
-     *
-     * @param configuration the hadoop configuration to build on top of
-     * @return Builder
-     */
-    public WithCompressionBuilder<K, V> withConfiguration(Configuration configuration) {
-      return new WithCompressionBuilder<>(keyClass, valueClass, outputPath, false, configuration);
+    @Override
+    public Buildable<K, V> withConfiguration(Configuration configuration) {
+      this.configuration = Objects.requireNonNull(configuration);
+      return this;
     }
 
-    /**
-     * Optional setter for compression
-     *
-     * @param compressionClass COMPRESS_CODEC class
-     * @param compressionType COMPRESS_TYPE value
-     * @return Builder
-     */
-    public SinkBuilder<K, V> withCompression(
+    @Override
+    public Buildable<K, V> withCompression(
         Class<? extends CompressionCodec> compressionClass,
         SequenceFile.CompressionType compressionType) {
-      return withConfiguration(null).withCompression(compressionClass, compressionType);
+      this.compressionClass = Objects.requireNonNull(compressionClass);
+      this.compressionType = Objects.requireNonNull(compressionType);
+      return this;
     }
 
+    @Override
+    public Buildable<K, V> withLazyOutputFormat() {
+      this.useLazyOutputFormat = true;
+      return this;
+    }
+
+    @Override
     public SequenceFileSink<K, V> build() {
-      return withConfiguration(null).withCompression(null, null).build();
+      final Configuration newConfiguration = wrap(configuration);
+
+      if (compressionClass != null && compressionType != null) {
+        newConfiguration.setBoolean(
+            FileOutputFormat.COMPRESS, true);
+        newConfiguration.set(
+            FileOutputFormat.COMPRESS_TYPE, compressionType.toString());
+        newConfiguration.setClass(
+            FileOutputFormat.COMPRESS_CODEC, compressionClass, CompressionCodec.class);
+      }
+
+      return new SequenceFileSink<>(
+          keyClass, valueClass, outputPath, newConfiguration, useLazyOutputFormat);
     }
   }
 
-  public static class WithConfigurationBuilder<K, V> {
+  public interface OutputPathBuildable<K, V> {
 
-    private final Class<K> keyClass;
-    private final Class<V> valueClass;
-    private final String outputPath;
-    private final boolean useLazyOutputFormat;
+    /**
+     * Mandatory output path.
+     *
+     * @param outputPath the destination where to save the output
+     * @return Builder with optional setters
+     */
+    Buildable<K, V> outputPath(String outputPath);
+  }
 
-    WithConfigurationBuilder(
-        Class<K> keyClass,
-        Class<V> valueType,
-        String outputPath,
-        boolean useLazyOutputFormat) {
-      this.keyClass = keyClass;
-      this.valueClass = valueType;
-      this.outputPath = outputPath;
-      this.useLazyOutputFormat = useLazyOutputFormat;
-    }
+  public interface Buildable<K, V> {
```

Comment:
Can we have a better name for this interface? For example `OptinalBuilder`? I also think that we should stick with `*Builder` naming convention for interfaces

Reply:
OK, I renamed `Buildable` to `OptionalBuilder` and `OutputPathBuildable` to `OutputPathBuilder`